### PR TITLE
[Woo POS] Implement the UI of the product list

### LINKED ISSUE #11785: [Woo POS] Implement UI of the cart items
Image

### LINKED ISSUE #11807: [Woo POS] Implement the UI of the product list

### PR DESCRIPTION
**Pull request by kidinov**
Closes: #11785 

### Description

* Adds the UI of the products list
* Adds reusable shimmer component 
* Adds loading state and "loading more" indicator
* Base for adding custom colors to the theme

We will still need to implement these states:
* Empty
* Error
* Add pull to refresh

### Testing information

* Point attention to the UI
* Try opening it when there are no products in the cache to validate
* Load more and notice loading more indicator

### Images/gif

https://github.com/woocommerce/woocommerce-android/assets/4923871/77174a67-60a8-4a02-8b5f-9cdeaec610cf

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>CustomColors</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsDataSource.kt [hunk 1]
```diff
@@ -1,10 +1,38 @@
...
 
-interface WooPosProductsDataSource {
-    val products: Flow<List<Product>>
-    suspend fun loadSimpleProducts()
-    suspend fun loadMore()
+@Singleton
+class WooPosProductsDataSource @Inject constructor(private val handler: ProductListHandler) {
+    var hasMorePages: AtomicBoolean = AtomicBoolean(false)
+
+    val products: Flow<List<Product>> = handler.productsFlow
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsDataSource.kt:17]: The API of the handler is really hard to use. They return errors as a result of method calling and products as a stream. I think we'll need to wrap it to something like:

```
suspend fun loadSimpleProducts(): Flow<Result<List>>
suspend fun loadMore(): Result<List>
```

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsDataSource.kt [hunk 2]
```diff
@@ -1,10 +1,38 @@
...
+        .onEach { hasMorePages.set(it.size % PAGE_SIZE == 0) }
```
Comment (samiuelson) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsDataSource.kt]: 💡 Have you considered `ProductListHandler.canLoadMore` publicly, and using it directly instead?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsDataSource.kt [hunk 3]
```diff
@@ -1,10 +1,38 @@
...
+    }
+
+    suspend fun loadMore() {
+        handler.loadMore()
+    }
+
+    private companion object {
+        // the product list handler doesn't expose page size
+        // so workaround to copy it here
+        const val PAGE_SIZE = 25
```
Comment (samiuelson) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsDataSource.kt]: 💡 I think it's better to make the actual `PAGE_SIZE` public in handler

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsDataSource.kt (hunk 1 above)
Comment (samiuelson) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsDataSource.kt:17]: What particular problem is hard to solve with the existing API?

**Review by samiuelson — APPROVED** [review]
Works good 👍🏼 I left some follow up questions, let me know what you think.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsDataSource.kt (hunk 1 above)
Comment (kidinov) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsDataSource.kt:17]: Let's say `loadSimpleProducts` returns an error, and `products` gets new data. Are we showing an error or products? How do we even know if this product is for our current request of previous and, let's say, the first failed and the second succeeded?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsDataSource.kt (hunk 2 above)
Comment (kidinov) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsDataSource.kt]: Good point - done here 843a6aec0bb4bfc761862c26f12f03b1fe01af00

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsDataSource.kt (hunk 3 above)
Comment (kidinov) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsDataSource.kt]: Improved by that suggestion - https://github.com/woocommerce/woocommerce-android/pull/11821#discussion_r1658524019

**Pull request merged**